src/lib.rs

```diff
@@ -61,69 +76,253 @@ impl InlinerBuilder {
         self
     }
 
-    /// Configures whether unexpanded modules (due to a missing file) will lead
-    /// to an `Err` return value or not.
+    /// Configures whether unexpanded modules (due to missing files or invalid Rust source code)
+    /// will lead to an `Err` return value or not.
     ///
     /// Default: `false`.
-    pub fn error_not_found(&mut self, error_not_found: bool) -> &mut Self {
-        self.error_not_found = error_not_found;
+    pub fn track_errors(&mut self, track_errors: bool) -> &mut Self {
+        self.track_errors = track_errors;
         self
     }
 
     /// Parse the source code in `src_file` and return a `syn::File` that has all modules
     /// recursively inlined.
-    pub fn parse_and_inline_modules(&self, src_file: &std::path::Path) -> Result<syn::File, Error> {
+    pub fn parse_and_inline_modules(&self, src_file: &Path) -> Result<syn::File, Error> {
         self.parse_internal(src_file, FsResolver::default())
     }
 
     fn parse_internal<R: FileResolver + Clone>(
         &self,
-        src_file: &std::path::Path,
+        src_file: &Path,
         resolver: R,
     ) -> Result<syn::File, Error> {
-        let mut errors = if self.error_not_found {
+        let mut errors = if self.track_errors {
             Some(vec![])
         } else {
             None
         };
         let result =
             Visitor::<R>::with_resolver(src_file, self.root, errors.as_mut(), Cow::Owned(resolver))
-                .visit();
+                .visit()
+                .map_err(Error::Root)?;
         match errors {
             Some(ref errors) if errors.is_empty() => Ok(result),
             None => Ok(result),
-            Some(errors) => Err(Error::NotFound(errors)),
+            Some(errors) => Err(Error::Inline(IncompleteInlining::new(result, errors))),
         }
     }
 }
 
-/// A source location
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub struct SourceLocation {
-    pub path: PathBuf,
-    pub line: usize,
-    pub column: usize,
-    _private: (),
+/// An error that was encountered while reading, parsing or inlining a module.
+#[derive(Debug)]
+pub enum Error {
+    /// An error happened while reading or parsing the initial file.
+    ///
+    /// For example, the initial file wasn't found or wasn't a valid Rust file.
+    Root(ErrorKind),
+
+    /// The initial file was successfully loaded, but some errors happened while attempting to
+    /// inline modules.
+    Inline(IncompleteInlining),
```

Comment:
I think returning the file makes a lot of sense! Is there any utility to having `IncompleteInlining` be a struct of its own? Or can it just be

```rust
Inline {
    output: syn::File,
    errors: Vec<InlineError>,
}

Reply:
I went with the struct here for two reasons, but neither one is set in stone:

1. I wanted to uphold the invariant that `errors` would never be empty. Doing that with an enum is only possible when using `#[non_exhaustive]`, and I'd prefer to maintain backwards compatibility to rustc 1.32 because that's one less thing to think about.
2. I have a slight personal bias against struct-style enum variants, mostly because there's no obvious way to pass them between functions without destructuring them and putting their members into a tuple. I've seen other projects in the past undertake sometimes-painful migrations from struct-style variants to newtype-style variants, and came away with the belief that struct-style variants should be used only in very narrow scenarios.

Reply:
Yeah I have the same bias for the same reasons too — I think in this case it might be justifiable though.

Reply:
I think [this thread](https://github.com/TedDriggs/syn-inline-mod/pull/13#discussion_r451850012) resolves the issue; we'll need the `InliningResult` struct, and we'll relax the "must have non-zero errors" requirement and remove the error tracking option.